internal/plugincommon/discovery/processor.go

```diff
@@ -0,0 +1,120 @@
+package discovery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/chainlink-ccip/internal/plugincommon"
+	dt "github.com/smartcontractkit/chainlink-ccip/internal/plugincommon/discovery/discoverytypes"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/consts"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/reader"
+)
+
+// Ensure that PluginProcessor is implemented.
+var _ plugincommon.PluginProcessor[dt.Query, dt.Observation, dt.Outcome] = &ContractDiscoveryProcessor{}
+
+// ContractDiscoveryProcessor is a plugin processor for discovering contracts.
+type ContractDiscoveryProcessor struct {
+	lggr      logger.Logger
+	reader    *reader.CCIPReader
+	homechain reader.HomeChain
+	dest      cciptypes.ChainSelector
+	bigF      int
```

Comment:
nit: this is now called `fRoleDON` in the contracts, I think its a more descriptive name :-)